docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,694 @@
+        subjectAltNames: # if subjectAltNames is not defined then take domain or ips
+          - example.com
+          - "spiffe://example.local/ns/local"
+        caCert: 
+          inline: 123
+        clientCert:
+          secret: 123
+        clientKey:
+          secret: 123
+status: 
+  vip:
+    value: 242.0.0.1
+    type: Kuma
+  addresses:
+  - hostname: example.ext.svc.local
+    status: NotAvailable
+    origin:
+      kind: HostnameGenerator
+      name: k8s-example-service-hostname
+    reason: "addresses are overlapping with ext2"
+```
+* **spec**:
+  * **match**: defines a traffic that should be routed through the sidecar
+    * **type**: type of the match, one of `KumaHostname`, `Domain`, `CIDR` and `IP` are available
+      * `KumaHostname`: matches hostnames provided by HostnameGenerator
+      * `Domain`: handles traffic to the specified domain
+      * `CIDR`: handles traffic to specified addresses range
+      * `IP` handles the traffic to specified IP
+    * **value**: depends of the type can be an existing domain, CIDR or IP. Not available when type `KumaHostname`
+    * **port**: defines a port to which a user does requests
+    * **protocol**: defines a protocol of the communication. Possible values:
+      * `tcp`: WARNING: shouldn't be used when match is `Domain` type. On the TCP level we are not able to disinguish domain, in this case it is going to hijack whole traffic on this port. We are going to validate configuration and do not apply config when protocol is tcp and `type: Domain`.
+      * `grpc`
+      * `http`
+      * `http2`
+  * **type**: defines what kind of destination it is, currently `Managed`
+    * `Managed`: allows creating a set of destination endpoints and `TLS` configuration
+  * **managed**: defines where matched requests should be routed, only for a `type: Managed`
+    * **extension**: struct for a plugin configuration
+      * **type**: defines what kind of plugin to use, it's a string type so any new plugins should works.
+      * **config**: json map that is mapped to configuration provided in the type.
+    * **endpoints**: defines a list of endpoints.
+      * **address**: defines an address to which a user want to send a request. Is possible to provide `domain`, `ip` and `unix` sockets
+      * **port**: defines a port of a destination.
+    * **tls**: provides a TLS configuration when proxy is resposible for a TLS origination
+      * **version**: section for providing version specification.
+        * **min**: defines minmum supported version. One of `TLS_AUTO`, `TLSv1_0`, `TLSv1_1`, `TLSv1_2`, `TLSv1_3`
+        * **max**: defines maximum supported version. One of `TLS_AUTO`, `TLSv1_0`, `TLSv1_1`, `TLSv1_2`, `TLSv1_3`
+      * **allowRenegotiation**: defines if TLS sessions will allow renegotiation.
+      * **verification**: section for providing TLS verification details.
+        * **skipSAN**: defines if proxy should skip SAN verification. Default `false`.
```

Comment:
Does it make sense to have a `skipVerification` (The insecure setup where the CA isn't verified) ?
If that's the case should it be `mode: SkipSAN, SkipCA, Secured, SkipALL` and default is `Secured`?

Reply:
> can we do Spiffe with trustDomain?

from spiffie docs:

> All workloads identified in the same trust domain are issued identity documents that can be verified against the root keys of the trust domain.

and for [SVID](https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md#2-spiffe-id):

> In an X.509 SVID, the corresponding SPIFFE ID is set as a URI type in the Subject Alternative Name extension (SAN extension, see [RFC 5280 section 4.2.16](https://tools.ietf.org/html/rfc5280#section-4.2.1.6)).

so do we need anything else than the same CA as trust domain in `caCert` and matching SAN in `subjectAltNames`?

Reply:
Except that the validation maybe a SAN prefix no?

Reply:
Yeah, we could extent the subjectAltNames section to be (there is a string matcher on the Envoy side):

```
        subjectAltNames: # if subjectAltNames is not defined then take domain or ips
          - type: Exact
             example.com
          - type: Prefix
             "spiffe://my-trust-domain"
```

WDYT?

Reply:
let's go for this and default type is `Exact`